src/librustc_passes/rvalue_promotion.rs

```diff
@@ -445,40 +485,141 @@ fn check_expr<'a, 'tcx>(v: &mut CheckCrateVisitor<'a, 'tcx>, e: &hir::Expr, node
             }
         }
 
-        hir::ExprField(ref expr, _) => {
+        hir::ExprField(ref expr, _ident) => {
+            match *expr {
+                ref expr => { v.visit_expr(&expr) },
+            }
             if let Some(def) = v.tables.expr_ty(expr).ty_adt_def() {
                 if def.is_union() {
                     v.promotable = false
                 }
             }
         }
 
-        hir::ExprBlock(..) |
-        hir::ExprIndex(..) |
-        hir::ExprArray(_) |
-        hir::ExprType(..) |
-        hir::ExprTup(..) => {}
+        hir::ExprBlock(ref box_block, ref _option_label) => {
```

Comment:
Aha! You also need to visit all statements inside the block, not just the `expr`

calling `visit_stmt` on them should suffice.